File: graph.py
class Graph:
    def __init__(self):
        self.count = 0
        self.adject_list = {}

    # adding a Node (new item)
    def add_vertex(self, node):
        self.adject_list[node] = []
        self.count += 1

    # undirected graph
    def add_edge(self, node1, node2):
        self.adject_list[node1].append(node2)
        self.adject_list[node2].append(node1)


def bfs(graph, start_node):
    queue = []
    visited = []
    res = []

    visited.append(start_node)
    queue.append(start_node)

    while len(queue):
        vertex = queue.pop(0)
        res.append(vertex)
        for neighbor in graph[vertex]:
            if neighbor not in visited:
                visited.append(neighbor)
                queue.append(neighbor)

    return res

def dfs(graph, node: str, visited: set, res):
    if not node:
        return

    visited.add(node)
    res.append(node)

    for neighbor in graph[node]:
        if neighbor not in visited:
            dfs(graph, neighbor, visited, res)

    return res
    

def main():
    graph = Graph()
    graph.add_vertex("1")
    graph.add_vertex("5")
    graph.add_vertex("6")
    graph.add_vertex("2")
    graph.add_edge("1", "5")
    graph.add_edge("1", "6")
    graph.add_edge("5", "2")
    graph.add_edge("6", "2")
    
    #BFS
    bfs(graph.adject_list, "1")

    #DFS
    visited = set()
    res = dfs(graph.adject_list, "5", visited, [])

    #BFS
    res2 = bfs(graph.adject_list, "1")
    print(res, res2)

if __name__ == "__main__":
    main()

File: graph.ts
interface adjacentList<T> {
    value: string
    count: number
    data: adjacentList<T>[];
}

const addVertex = <T>(node: string, graph: adjacentList<T>) => {
    graph.data[node] = [];
    graph.count++;
    return;
}

const addEdge = <T>(node1: string, node2: string, graph: adjacentList<T>) => {
    graph.data[node1].push(node2);
    graph.data[node2].push(node1);
    return;
}


// Iterative way to do DFS
const showConnections = <T>(list: adjacentList<T>) => {
    const allNodes: string[] = Object.keys(list.data);
    for (let node of allNodes) {
        let nodeConnections: string[] = list.data[node];
        let connections = "";
        let vertex: string;
        for (vertex of nodeConnections) {
            connections += vertex + " ";
        }
        console.log(node + "-->" + connections);
    }
}

const dfs = <T>(graph: adjacentList<T>, node: T, res: T[], visited): Array<T> => {
    if (!node)
        return res;

    visited[node.toString()] = 1;
    let i: T;
    for (i of graph.data[node.toString()])
    {
        if (!visited[i.toString()])
        {
            visited[i.toString()] = 1;
            res.push(i);
            dfs(graph, i, res, visited);
        }
    }
    return res;
}

const bfs = <T>(graph: adjacentList<T>, node: T): T[] => {
    let queue: T[] = [];
    let visited = {};
    let res: T[] = [];
    queue.push(node);
    while (queue.length > 0)
    {
        const vertex: T = queue.shift();
        let i: T;
        for (i of graph.data[vertex.toString()])
        {
            if (!visited[i.toString()])
            {
                queue.push(i);
                res.push(i);
                visited[i.toString()] = i;
            }
        }
    }
    return res;
}

function main() {
    let graph: adjacentList<string> = { value: "", count: 0, data: [] };
    addVertex("1", graph);
    addVertex("2", graph);
    addVertex("5", graph);
    addVertex("3", graph);
    addVertex("4", graph);
    addEdge<string>("3", "2", graph);
    addEdge<string>("4", "5", graph);
    addEdge<string>("4", "2", graph);
    //showConnections<string>(graph);
    const res = dfs<string>(graph, "4", [], {});
    const res2 = bfs<string>(graph, "4");
    console.log(res, res2);
}

main();
